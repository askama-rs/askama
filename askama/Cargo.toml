[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
