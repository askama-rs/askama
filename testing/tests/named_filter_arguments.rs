use askama::Template;

#[test]
fn test_pluralize_two_positional_args() {
    #[derive(Template)]
    #[template(
        source = r#"I have {{ count }} butterfl{{ count | pluralize("y", "ies") }}."#,
        ext = "txt"
    )]
    struct CountButterflies {
        count: usize,
    }

    assert_eq!(
        CountButterflies { count: 0 }.render().unwrap(),
        "I have 0 butterflies."
    );
    assert_eq!(
        CountButterflies { count: 1 }.render().unwrap(),
        "I have 1 butterfly."
    );
    assert_eq!(
        CountButterflies { count: 2 }.render().unwrap(),
        "I have 2 butterflies."
    );
}

#[test]
fn test_pluralize_positional_and_named() {
    #[derive(Template)]
    #[template(
        source = r#"I have {{ count }} butterfl{{ count | pluralize("y", pl = "ies") }}."#,
        ext = "txt"
    )]
    struct CountButterflies {
        count: usize,
    }

    assert_eq!(
        CountButterflies { count: 0 }.render().unwrap(),
        "I have 0 butterflies."
    );
    assert_eq!(
        CountButterflies { count: 1 }.render().unwrap(),
        "I have 1 butterfly."
    );
    assert_eq!(
        CountButterflies { count: 2 }.render().unwrap(),
        "I have 2 butterflies."
    );
}

#[test]
fn test_pluralize_named_reordered() {
    #[derive(Template)]
    #[template(
        source = r#"I have {{ count }} butterfl{{ count | pluralize(pl = "ies", sg = "y") }}."#,
        ext = "txt"
    )]
    struct CountButterflies {
        count: usize,
    }

    assert_eq!(
        CountButterflies { count: 0 }.render().unwrap(),
        "I have 0 butterflies."
    );
    assert_eq!(
        CountButterflies { count: 1 }.render().unwrap(),
        "I have 1 butterfly."
    );
    assert_eq!(
        CountButterflies { count: 2 }.render().unwrap(),
        "I have 2 butterflies."
    );
}

#[test]
fn test_pluralize_defaulted_and_named() {
    #[derive(Template)]
    #[template(
        source = r#"I have {{ count }} potato{{ count | pluralize(pl = "es") }}."#,
        ext = "txt"
    )]
    struct CountPotatoes {
        count: usize,
    }

    assert_eq!(
        CountPotatoes { count: 0 }.render().unwrap(),
        "I have 0 potatoes."
    );
    assert_eq!(
        CountPotatoes { count: 1 }.render().unwrap(),
        "I have 1 potato."
    );
    assert_eq!(
        CountPotatoes { count: 2 }.render().unwrap(),
        "I have 2 potatoes."
    );
}

#[test]
fn test_ident() {
    const TEXT: &str = "\
        Lorem Ipsum has been the industry's\n\
        \n\
        standard dummy text ever since the 1500s";

    #[derive(Template)]
    #[template(source = r#"<{{ text | indent("$$",  blank = true) }}>"#, ext = "txt")]
    struct IdentBlank<'a> {
        text: &'a str,
    }

    assert_eq!(
        IdentBlank { text: TEXT }.render().unwrap(),
        "\
            <Lorem Ipsum has been the industry's\n\
            $$\n\
            $$standard dummy text ever since the 1500s>"
    );

    #[derive(Template)]
    #[template(source = r#"<{{ text | indent("$$",  first = true) }}>"#, ext = "txt")]
    struct IdentFirst<'a> {
        text: &'a str,
    }

    assert_eq!(
        IdentFirst { text: TEXT }.render().unwrap(),
        "\
            <$$Lorem Ipsum has been the industry's\n\
            \n\
            $$standard dummy text ever since the 1500s>"
    );

    #[derive(Template)]
    #[template(
        source = r#"<{{ text | indent("$$", blank = true, first = true) }}>"#,
        ext = "txt"
    )]
    struct IdentBoth<'a> {
        text: &'a str,
    }

    assert_eq!(
        IdentBoth { text: TEXT }.render().unwrap(),
        "\
            <$$Lorem Ipsum has been the industry's\n\
            $$\n\
            $$standard dummy text ever since the 1500s>"
    );

    #[derive(Template)]
    #[template(
        source = r#"<{{ text | indent("$$", first = true, blank = true) }}>"#,
        ext = "txt"
    )]
    struct IdentBoth2<'a> {
        text: &'a str,
    }

    assert_eq!(
        IdentBoth2 { text: TEXT }.render().unwrap(),
        "\
            <$$Lorem Ipsum has been the industry's\n\
            $$\n\
            $$standard dummy text ever since the 1500s>"
    );

    #[derive(Template)]
    #[template(
        source = r#"<{{ text | indent(first = true, width = "$$", blank = true) }}>"#,
        ext = "txt"
    )]
    struct IdentBoth3<'a> {
        text: &'a str,
    }

    assert_eq!(
        IdentBoth3 { text: TEXT }.render().unwrap(),
        "\
            <$$Lorem Ipsum has been the industry's\n\
            $$\n\
            $$standard dummy text ever since the 1500s>"
    );
}
