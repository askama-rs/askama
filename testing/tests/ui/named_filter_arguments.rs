use askama::Template;

#[derive(Template)]
#[template(
    source = r#"I have {{ count }} butterfl{{ count | pluralize(pl = "ies", "y") }}."#,
    ext = "txt"
)]
struct PositionalAfterNamed {
    count: usize,
}

#[derive(Template)]
#[template(
    source = r#"I have {{ count }} butterfl{{ count | pluralize(pl = "y", pl = "ies") }}."#,
    ext = "txt"
)]
struct NamedRepeated {
    count: usize,
}

#[derive(Template)]
#[template(
    source = r#"I have {{ count }} butterfl{{ count | pluralize("y", sg = "ies") }}."#,
    ext = "txt"
)]
struct NamedButAlreadyPositional {
    count: usize,
}

#[derive(Template)]
#[template(
    source = r#"I have {{ count }} butterfl{{ count | pluralize("y", plural = "ies") }}."#,
    ext = "txt"
)]
struct NoSuchNamedArgument {
    count: usize,
}

#[derive(Template)]
#[template(
    source = r#"Scream: {{ message | uppercase(case = "upper") }}"#,
    ext = "txt"
)]
struct NamedArgumentButNoArgumentExpected<'a> {
    message: &'a str,
}

fn main() {}
