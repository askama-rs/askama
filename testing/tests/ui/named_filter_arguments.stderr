error: named arguments must always be passed last
 --> <source attribute>:1:47
       "(pl = \"ies\", \"y\") }}."
 --> tests/ui/named_filter_arguments.rs:5:14
  |
5 |     source = r#"I have {{ count }} butterfl{{ count | pluralize(pl = "ies", "y") }}."#,
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: named argument `pl` was passed more than once
 --> <source attribute>:1:47
       "(pl = \"y\", pl = \"ies\") }}."
  --> tests/ui/named_filter_arguments.rs:14:14
   |
14 |     source = r#"I have {{ count }} butterfl{{ count | pluralize(pl = "y", pl = "ies") }}."#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `sg` argument to `pluralize` filter was already set; its arguments are: ([sg], [pl])
 --> NamedButAlreadyPositional.txt:1:47
       "(\"y\", sg = \"ies\") }}."
  --> tests/ui/named_filter_arguments.rs:23:14
   |
23 |     source = r#"I have {{ count }} butterfl{{ count | pluralize("y", sg = "ies") }}."#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `pluralize` filter does not have an argument `plural`; its arguments are: ([sg], [pl])
 --> NoSuchNamedArgument.txt:1:47
       "(\"y\", plural = \"ies\") }}."
  --> tests/ui/named_filter_arguments.rs:32:14
   |
32 |     source = r#"I have {{ count }} butterfl{{ count | pluralize("y", plural = "ies") }}."#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `uppercase` filter does not have any arguments
 --> NamedArgumentButNoArgumentExpected.txt:1:30
       "(case = \"upper\") }}"
  --> tests/ui/named_filter_arguments.rs:41:14
   |
41 |     source = r#"Scream: {{ message | uppercase(case = "upper") }}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
